c/tskit/tables.c

```diff
@@ -4343,9 +4343,12 @@ cmp_mutation(const void *a, const void *b)
     const tsk_mutation_t *ib = (const tsk_mutation_t *) b;
     /* Compare mutations by site */
     int ret = (ia->site > ib->site) - (ia->site < ib->site);
+    /* Within a particular site sort by time if known, then ID. This ensures that
+     * relative ordering within a site is maintained */
+    if (ret == 0 && !tsk_is_unknown_time(ia->time) && !tsk_is_unknown_time(ib->time)) {
```

Comment:
I can't see how this wouldn't completely determine the sort order - if either time is unknown we should fall back on comparing by ID.

Reply:
The problem here is that our `cmp` gives inconsistent orderings (5<4, 4<U, U<5) in my example above as one is by time  and the other two by id. The cycle has to be broken somehow.

Reply:
@petrelharp I think you might be helpful here as I'm not sure how to fix this one! This comment hopefully explains the situation: https://github.com/tskit-dev/tskit/pull/726#issuecomment-662429215